### PR TITLE
[Events Orderbook] Refactor `applyEvent` Logic

### PR DESCRIPTION
**Pull request by nlordell**
The refactor that was promised! This PR refactors `applyEvents` switch statement to be more "dumb" by forwarding each event to their individual functions.

This has the added benefit of using the `Event` types so that we have type safety on the event `returnValues` property.

### Test Plan

This PR is just a refactor so unit tests should verify that nothing broke.

### CODE REVIEW

**Review by e00E — APPROVED** [review]

File: src/streamed/state.ts [hunk 1]
```diff
@@ -165,135 +167,256 @@ export class AuctionState {
...
+
+    this.account(user).pendingWithdrawals.delete(tokenAddr)
+  }
+
+  /**
+   * Applies a withdraw request event to the auction state.
+   *
+   * @throws
+   * In strict mode, throws if the withdraw request is placed in the over top of
+   * and existing unapplied request.
```
Comment (bh2smith) [inline src/streamed/state.ts]: I think we have some typos here. Also, I am not sure why a withdraw request placed over top of an existing should throw. I believe throw contract would overwrite the existing one.

**Review by bh2smith — APPROVED** [review]
Lots of good stuff here. Just a typo detected on the withdraw request doc string. The entire sentence is off, just read it again and it should be pretty easy to see the correction.

File: src/streamed/state.ts (hunk 1 above)
Comment (nlordell) [reply src/streamed/state.ts]: :see_no_evil: 

In the smart contract, a withdraw is done to clear the previous withdraw request if it has already matured. If we ever get a withdraw request that would effectively overwrite a previous one, it means that we missed a withdraw event somewhere.

**Review by fleupold — COMMENTED** [review]
There seems to be more going on than just refactoring the switch statements in the 300 added lines (I also see some new unit tests). E.g. I don't understand why tokens was changed from a map to an array?

**Comment by nlordell** [comment]
>  E.g. I don't understand why tokens was changed from a map to an array?

You're right, I'll split that into a separate PR. 

> There seems to be more going on than just refactoring the switch statements in the 300 added lines

Besides the tokens map => array refactor, there anything else. We use strong typing for the events but that ties in to the refactor and is mentioned in the PR description.

> (I also see some new unit tests) 

The new unit test is for a new token ID assertion that was added when I converted the tokens map into an array

Also the deposit unit test was refactored into 2 smaller tests.

**Comment by fleupold** [comment]
> Besides the tokens map => array refactor, there anything else

I'm not sure. I was expecting a relatively easy to follow change where we just move some code into functions (which is why it might be relatively large) but then saw some other stuff that made it hard to review for me. But given that others have reviewed and accepted it feel free to merge without splitting it.

**Comment by nlordell** [comment]
> But given that others have reviewed and accepted it feel free to merge without splitting it.

I think it makes sense to split it here. I was just being a bit lazy :stuck_out_tongue: 

> I was expecting a relatively easy to follow change where we just move some code into functions

I think its because I had some switch branches that already had their own function (like `addOrder`, which got renamed to `applyOrderPlacement`) and others where the logic was in the switch branch (like `Trade` event for example).

Also, using strict types (I didn't want to use `any` everywhere) changed the code a bit, in `src/streamed/index.ts` for example, where we have a `getPastEvents` function now so that we only need to cast to `AllEvents<BatchExchange>` in one place.

I think there is value in trying to split this up so its easier to review. Also, I completely agree that the diff view here makes it look way worse than it actually is, and makes the changes a bit harder to follow.

**Comment by nlordell** [comment]
@fleupold I created #739 with the token map => array refactor, let me know if its a bit easier to review.

Unfortunately because some code got moved out of the switch statement, and the order in which it code blocks appear has changed, it makes seeing the actual diff still hard :( I'm not sure how to make this easier to review, but I'm open to suggestions.

File: src/streamed/state.ts [hunk 2]
```diff
@@ -165,135 +167,258 @@ export class AuctionState {
...
-    /* eslint-enable no-case-declarations */
+  }
+
+  /**
+   * Applies a deposit event to the auction state.
+   */
+  private applyDeposit(
+    { user, token, amount: depositAmount }: Event<BatchExchange, "Deposit">,
+  ) {
+    this.updateBalance(user, token, amount => amount + BigInt(depositAmount))
```
Comment (nlordell) [inline src/streamed/state.ts:226]: This code was moved from here:
https://github.com/gnosis/dex-contracts/pull/725/files#diff-4b1723270e0b82c809346fdae9c8d383L179-L183

Comment (nlordell) [inline src/streamed/state.ts:226]: https://github.com/gnosis/dex-contracts/blob/089466be72b96b943ea5230a6b104b84fe5eaa24/src/streamed/state.ts#L179-L183

File: src/streamed/state.ts [hunk 3]
```diff
@@ -165,135 +167,258 @@ export class AuctionState {
...
+  }
+
+  /**
+   * Applies an order cancellation event to the auction state.
+   */
+  private applyOrderCancellation({ owner, id }: Event<BatchExchange, "OrderCancellation">) {
+    // TODO(nlordell): We need to pre-fetch the batch ID based on the block
+    // number for this event to be able to accurately set this value.
+    this.order(owner, parseInt(id)).validUntil = null
```
Comment (nlordell) [inline src/streamed/state.ts:235]: https://github.com/gnosis/dex-contracts/blob/089466be72b96b943ea5230a6b104b84fe5eaa24/src/streamed/state.ts#L185-L187

File: src/streamed/state.ts [hunk 4]
```diff
@@ -165,135 +167,258 @@ export class AuctionState {
...
+  private applyOrderDeletion({ owner, id }: Event<BatchExchange, "OrderDeletion">) {
+    const order = this.order(owner, parseInt(id))
+    order.buyToken = 0
+    order.sellToken = 0
+    order.validFrom = 0
+    order.validUntil = 0
+    order.priceNumerator = BigInt(0)
+    order.priceDenominator = BigInt(0)
+    order.remainingAmount = BigInt(0)
+  }
```
Comment (nlordell) [inline src/streamed/state.ts]: https://github.com/gnosis/dex-contracts/blob/089466be72b96b943ea5230a6b104b84fe5eaa24/src/streamed/state.ts#L430-L439

File: src/streamed/state.ts [hunk 5]
```diff
@@ -165,135 +167,258 @@ export class AuctionState {
...
+
+    userOrders.push({
+      buyToken: parseInt(order.buyToken),
+      sellToken: parseInt(order.sellToken),
+      validFrom: parseInt(order.validFrom),
+      validUntil: parseInt(order.validUntil),
+      priceNumerator: BigInt(order.priceNumerator),
+      priceDenominator: BigInt(order.priceDenominator),
+      remainingAmount: BigInt(order.priceDenominator),
+    })
```
Comment (nlordell) [inline src/streamed/state.ts]: https://github.com/gnosis/dex-contracts/blob/089466be72b96b943ea5230a6b104b84fe5eaa24/src/streamed/state.ts#L191-L207

and

https://github.com/gnosis/dex-contracts/blob/089466be72b96b943ea5230a6b104b84fe5eaa24/src/streamed/state.ts#L385-L392

File: src/streamed/state.ts [hunk 6]
```diff
@@ -165,135 +167,258 @@ export class AuctionState {
...
   }
 
   /**
-   * Adds a token to the account state.
+   * Applies an emulated solution reversion event.
+   */
+  private applySolutionReversion(
+    { submitter, burntFees }: Event<BatchExchange, "SolutionSubmission">,
+  ) {
+    this.updateBalance(submitter, 0, amount => amount - BigInt(burntFees))
```
Comment (nlordell) [inline src/streamed/state.ts:286]: https://github.com/gnosis/dex-contracts/blob/089466be72b96b943ea5230a6b104b84fe5eaa24/src/streamed/state.ts#L237-L239

File: src/streamed/state.ts [hunk 7]
```diff
@@ -165,135 +167,258 @@ export class AuctionState {
...
+    { submitter, burntFees }: Event<BatchExchange, "SolutionSubmission">,
+  ) {
+    this.updateBalance(submitter, 0, amount => amount + BigInt(burntFees))
+    if (this.options.strict) {
+      for (const [user, { balances }] of this.accounts.entries()) {
+        for (const [token, balance] of balances.entries()) {
+          assert(balance >= BigInt(0), `user ${user} token ${token} balance is negative`)
+        }
+      }
+    }
```
Comment (nlordell) [inline src/streamed/state.ts:308]: https://github.com/gnosis/dex-contracts/blob/089466be72b96b943ea5230a6b104b84fe5eaa24/src/streamed/state.ts#L209-L212

and

https://github.com/gnosis/dex-contracts/blob/089466be72b96b943ea5230a6b104b84fe5eaa24/src/streamed/state.ts#L478-L485

File: src/streamed/state.ts [hunk 8]
```diff
@@ -165,135 +167,258 @@ export class AuctionState {
...
+    this.updateOrderRemainingAmount(
+      trade.owner,
+      parseInt(trade.orderId),
+      amount => amount - BigInt(trade.executedSellAmount),
+    )
+    this.updateBalance(
+      trade.owner,
+      parseInt(trade.buyToken),
+      amount => amount + BigInt(trade.executedBuyAmount),
+    )
```
Comment (nlordell) [inline src/streamed/state.ts:348]: https://github.com/gnosis/dex-contracts/blob/089466be72b96b943ea5230a6b104b84fe5eaa24/src/streamed/state.ts#L217-L234

File: src/streamed/state.ts [hunk 9]
```diff
@@ -165,135 +167,258 @@ export class AuctionState {
...
+      trade.owner,
+      parseInt(trade.orderId),
+      amount => amount + BigInt(trade.executedSellAmount),
+    )
+    this.updateBalance(
+      trade.owner,
+      parseInt(trade.buyToken),
+      amount => amount - BigInt(trade.executedBuyAmount),
+    )
+  }
```
Comment (nlordell) [inline src/streamed/state.ts]: https://github.com/gnosis/dex-contracts/blob/089466be72b96b943ea5230a6b104b84fe5eaa24/src/streamed/state.ts#L235-L256

File: src/streamed/state.ts [hunk 10]
```diff
@@ -165,135 +167,258 @@ export class AuctionState {
...
+    )
+
+    if (this.options.strict) {
+      assert(
+        newBalance >= BigInt(0),
+        `overdrew user ${user} token ${token} balance`,
+      )
+    }
+
+    this.account(user).pendingWithdrawals.delete(tokenAddr)
```
Comment (nlordell) [inline src/streamed/state.ts]: https://github.com/gnosis/dex-contracts/blob/089466be72b96b943ea5230a6b104b84fe5eaa24/src/streamed/state.ts#L265-L278

and

https://github.com/gnosis/dex-contracts/blob/089466be72b96b943ea5230a6b104b84fe5eaa24/src/streamed/state.ts#L374-L383

File: src/streamed/state.ts [hunk 11]
```diff
@@ -165,135 +167,258 @@ export class AuctionState {
...
+      assert(
+        existingBatch === undefined || existingBatch === batch,
+        `pending withdrawal for user ${user} modified from batch ${existingBatch} to ${batchId}`,
+      )
+    }
+
+    this.account(user).pendingWithdrawals.set(tokenAddr, {
+      batchId: batch,
+      amount: BigInt(amount),
+    })
```
Comment (nlordell) [inline src/streamed/state.ts:421]: https://github.com/gnosis/dex-contracts/blob/089466be72b96b943ea5230a6b104b84fe5eaa24/src/streamed/state.ts#L357-L372

**Review by nlordell — COMMENTED** [review]
@fleupold I added some links to the current code so its easier to see how the code changed and where it came from. Let me know if this helps!

**Review by fleupold — APPROVED** [review]
Thanks for splitting and adding the crosslinks! I was able to trace the moved code before (no need to do that step next time as I envision it being quite time consuming and annoying)

**Comment by bh2smith** [comment]
> But given that others have reviewed and accepted it feel free to merge without splitting it.

Right, I probably should have included a `TL;DR` on my approval.

**Pull request merged**